internal/controller/maintenancereplica_controller.go

```diff
@@ -0,0 +1,217 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and IronCore contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package controller
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/go-logr/logr"
+	"github.com/ironcore-dev/controller-utils/clientutils"
+	metalv1alpha1 "github.com/ironcore-dev/metal-operator/api/v1alpha1"
+)
+
+const (
+	ServerMaintenanceReplicaFinalizer = "metal.ironcore.dev/ServerMaintenanceReplica"
+)
+
+// ServerMaintenanceReplicaReconciler reconciles a ServerMaintenanceReplica object
+type ServerMaintenanceReplicaReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=metal.ironcore.dev,resources=ServerMaintenanceReplicas,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=metal.ironcore.dev,resources=ServerMaintenanceReplicas/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=metal.ironcore.dev,resources=ServerMaintenanceReplicas/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the ServerMaintenanceReplica object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile
+func (r *ServerMaintenanceReplicaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	replica := &metalv1alpha1.ServerMaintenanceReplica{}
+
+	if err := r.Get(ctx, req.NamespacedName, replica); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.reconcileExists(ctx, replica)
+}
+
+func (r *ServerMaintenanceReplicaReconciler) reconcileExists(ctx context.Context, replica *metalv1alpha1.ServerMaintenanceReplica) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	if !replica.DeletionTimestamp.IsZero() {
+		return r.delete(ctx, log, replica)
+	}
+
+	if modified, err := clientutils.PatchEnsureFinalizer(ctx, r.Client, replica, ServerMaintenanceReplicaFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("Ensured finalizer has been added")
+
+	return r.reconcile(ctx, log, replica)
+}
+
+func (r *ServerMaintenanceReplicaReconciler) delete(ctx context.Context, log logr.Logger, replica *metalv1alpha1.ServerMaintenanceReplica) (ctrl.Result, error) {
+	log.V(1).Info("Deleting ServerMaintenanceReplica")
```

Comment:
I think deleting a `ServerMaintenanceSet` should at least deleted the `Pending` maintenances it owns. What do you think?